src/clients/FtpClient.cc

```diff
@@ -563,17 +618,24 @@ Ftp::Client::handleEpsvReply(Ip::Address &remoteAddr)
     /*  ANSI sez [^0-9] is undefined, it breaks on Watcom cc */
     debugs(9, 5, "scanning: " << ctrl.last_reply);
 
-    buf = ctrl.last_reply + strcspn(ctrl.last_reply, "(");
+    const auto line = ctrl.last_reply;
+    const auto open = strchr(line, '(');
+    if (!open) {
+        debugs(9, DBG_IMPORTANT, "ERROR: Missing EPSV reply from " <<
+               ctrl.conn->remote << ": " <<
+               ctrl.last_reply);
+        return sendPassive();
+    }
 
-    char h1, h2, h3, h4;
-    unsigned short port;
-    int n = sscanf(buf, "(%c%c%c%hu%c)", &h1, &h2, &h3, &port, &h4);
+    SBuf tuple;
+    tuple.assign(open, strlen(open));
 
-    if (n < 4 || h1 != h2 || h1 != h3 || h1 != h4) {
+    Parser::Tokenizer tok(tuple);
+    uint16_t port = 0;
+    if (!parseEPSV(tok, tuple, port)) {
         debugs(9, DBG_IMPORTANT, "ERROR: Invalid EPSV reply from " <<
```

Comment:
This should have been level-2 as well. Please take the opportunity to fix that.
```suggestion
    if (!parseEPSV(tok, tuple, port)) {
        debugs(9, 2, "ERROR: Invalid EPSV reply from " <<
```